[AMDGPU][NFC][SIInsertWaitcnts] Remove redundant checks for invalidate instructions

stepthomas opened the pull request:
SIInsertWaitcnts::getVmemWaitEventType() tests for GLOBAL_INV, GLOBAL_WB, and GLOBAL_WBINV instructions, but in each case it is used, then either a check has already been made for these instructions, or it is known that the instruction definitely will not be one of these.

Move the checks for these instructions out of getVmemWaitEventType() into a new function getInvOrWBWaitEventType(), that returns an optional WaitEventType value, and use that in the situation where it could be one of the instructions.

SIInstrInfo::isGFX12CacheInvOrWBInst() is now itself redundant and is removed.

stepthomas committed:
remove redundant checks for invalidate instructions

SIInsertWaitcnts::getVmemWaitEventType() tests for GLOBAL_INV, GLOBAL_WB, and
GLOBAL_WBINV instructions, but in each case it is used, then either a
check has already been made for these instructions, or it is known that
the instruction definitely will not be one of these.

Move the checks for these instructions out of getVmemWaitEventType() into
a new function getInvOrWBWaitEventType(), that returns an optional WaitEventType
value, and use that in the situation where it could be one of the instructions.

SIInstrInfo::isGFX12CacheInvOrWBInst() is now itself redundant and is removed.

jmmartinez commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -527,19 +527,24 @@ class SIInsertWaitcnts {
-      break;
+      return {};
     }
+  }
```
You could move this into a static helper outside of the class now. It's not accessing any class members.

jmmartinez reviewed (commented):
Small comment. Please add some tags to the PR title: `[NFC][SIInsertWaitcnts]`

stepthomas committed:
Move getInvOrWBWaitEventType() out of SIInsertWaitcnts class

stepthomas committed:
Merge branch 'main' into stepthomas-AMDGPU-remove-redundant-waitcnt-check

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -530,16 +544,6 @@ class SIInsertWaitcnts {
   // Return the appropriate VMEM_*_ACCESS type for Inst, which must be a VMEM
   // instruction.
   WaitEventType getVmemWaitEventType(const MachineInstr &Inst) const {
-    switch (Inst.getOpcode()) {
```
Is there some reason why this function _should not_ handle these instructions? The only documented precondition is that Inst should be a VMEM instruction, which these are.

jayfoad replied:
I don't really get the motiviation. Isn't it clearer to keep the purpose of the function as simple as possible, and have it handle all VMEM instructions without exceptions?

If there _are_ some exceptions we should probably have an assert for them, but that would probably make the code at least as complex as not having the exceptions in the first place.

stepthomas replied:
The primary motivation was the observation that line 2268 ends up checking for GLOBAL_INV, GLOBAL_WB or GLOBAL_WBINV twice, the first time to check if it is one of those instructions, and the second time to work out which wait counter should be used. I wanted to streamline those two checks into a single check. The secondary motivation is the observation that the remaining uses of getVmemWaitEventType() could not be invalidate/WB instructions, because either they had already been dealt with or excluded by previous checks, so checking for them again is silly.

If it's felt that getVmemWaitEventType() should be fully general purpose, it could use getInvOrWBWaitEventType() to check for those specific instructions, but as of this patch, the remaining use cases mean it would be a redundant check.

jayfoad replied:
I'd prefer to keep getVmemWaitEventType general and instead look at why 2268 handles invalidate instructions specially. I think it could probably use the more general VMEM handling case at line 2277 except that it probably does not want to set IsVMEMAccess, because I think that really means "is this a VMEM access that increments Xcnt", and I guess invalidates do not increment Xcnt because they do not specify any address to translate.

stepthomas committed:
Update getVmemWaitEventType()'s description comment

stepthomas committed:
Merge branch 'main' into stepthomas-AMDGPU-remove-redundant-waitcnt-check

stepthomas committed:
Merge branch 'main' into stepthomas-AMDGPU-remove-redundant-waitcnt-check

stepthomas committed:
Address review comments.

Fold the logic that checks for invalidate/writeback instructions back into
getVmemWaitEventType(), but refactor the changes in updateEventWaitcntAfter()
so it now only calls it once for FLAT instructions and the right things
will happen.

This required some refactoring of mayAccessVMEMThroughFlat() so that it
no longer checks SIInstrInfo::usesVM_CNT() (users must now check that
themselves) and in the absence of memory operands it only returns true
if the instruction loads or stores. This allows the invalidate/writeback
instructions to not be considered as VMEM accesses.

stepthomas committed:
Merge branch 'main' into stepthomas-AMDGPU-remove-redundant-waitcnt-check
